icu4c/source/i18n/currunit.cpp

```diff
@@ -28,9 +28,13 @@ CurrencyUnit::CurrencyUnit(ConstChar16Ptr _isoCode, UErrorCode& ec) {
     // Note: in ICU4J Currency.getInstance(), we check string length for 3, but in ICU4C we allow a
     // non-NUL-terminated string to be passed as an argument, so it is not possible to check length.
     // However, we allow a NUL-terminated empty string, which should have the same behavior as nullptr.
+    // Consider NUL-terminated strings of length 1 or 2 as invalid.
     const char16_t* isoCodeToUse;
     if (U_FAILURE(ec) || _isoCode == nullptr || _isoCode[0] == 0) {
         isoCodeToUse = kDefaultCurrency;
+    } else if (_isoCode[1] == 0 || _isoCode[2] == 0) {
+        isoCodeToUse = _isoCode;
```

Comment:
It looks like error handling sets isoCodeToUse = kDefaultCurrency; so we should do so here, too (as I suggested in my ticket comment).

Optional: With that, other than the TODO comment, you could merge this with the not-invariant check below.

Reply:
> It looks like error handling sets isoCodeToUse = kDefaultCurrency; so we should do so here, too (as I suggested in my ticket comment).

Done; good catch.

> Optional: With that, other than the TODO comment, you could merge this with the not-invariant check below.

I made the error codes different.  This one is U_ILLEGAL_ARGUMENT_ERROR, and the other case uses U_INVARIANT_CONVERSION_ERROR.  So it's cleaner to have them be separate cases of the if statement.